cpp/src/parquet/encoding.cc

```diff
@@ -154,38 +146,42 @@ void AssertBinary(const arrow::Array& values) {
   }
 }
 
-template <typename EncoderType>
-void PutBinaryArray(const arrow::Array& values, EncoderType* encoder) {
+template <>
+void PlainEncoder<ByteArrayType>::Put(const arrow::Array& values) {
   AssertBinary(values);
   const auto& data = checked_cast<const arrow::BinaryArray&>(values);
+  const int64_t total_bytes = data.value_offset(data.length()) - data.value_offset(0);
+  PARQUET_THROW_NOT_OK(sink_.Reserve(total_bytes + data.length() * sizeof(uint32_t)));
+
+  auto AppendValue = [&](int64_t i) {
+    auto view = data.GetView(i);
+    const uint32_t length = static_cast<uint32_t>(view.size());
+    sink_.UnsafeAppend(&length, sizeof(uint32_t));
+    sink_.UnsafeAppend(view.data(), view.size());
+  };
+
   if (data.null_count() == 0) {
     // no nulls, just dump the data
     for (int64_t i = 0; i < data.length(); i++) {
-      encoder->Put(ByteArray(data.GetView(i)));
+      AppendValue(i);
     }
   } else {
     for (int64_t i = 0; i < data.length(); i++) {
       if (data.IsValid(i)) {
-        encoder->Put(ByteArray(data.GetView(i)));
+        AppendValue(i);
       }
     }
   }
 }
 
-template <>
-void PlainEncoder<ByteArrayType>::Put(const arrow::Array& values) {
-  PutBinaryArray(values, this);
-}
-
 template <>
 inline void PlainEncoder<FLBAType>::Put(const FixedLenByteArray* src, int num_values) {
   for (int i = 0; i < num_values; ++i) {
     // Write the result to the output stream
     if (descr_->type_length() > 0) {
       DCHECK(nullptr != src[i].ptr) << "Value ptr cannot be NULL";
```

Comment:
Above a similar assertion checks whether the length is 0 before asserting the pointer is non null

Reply:
I'm changing this to bail out altogether if type_length() is 0